File: 12-linked-list-cycle/12.js
import { logObject } from '../shared/utils.js';

class ListNode {
    constructor(val, next = null) {
        this.val = val;
        this.next = next;
    }
}

/**
/**
 * @param {ListNode} head
 * @return {boolean}
 */
var hasCycle = function (head) {
    let pos = head;
    let set = new Set([]);
    while (pos * pos.next) {
        logObject(set);
        if (set.has(pos)) {
            return true;
        }
        set.add(pos);
        pos = pos.next;
    }
    return false;
};

const l4 = new ListNode(-4);
const l2 = new ListNode(2, new ListNode(0, l4));
l4.next = l2;
logObject(hasCycle(new ListNode(3, l2)));

File: 12-linked-list-cycle/12.py
from typing import Optional

class ListNode:
    def __init__(self, x, nnext = None):
        self.val = x
        self.next = nnext

class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        pos = head
        sset = set()
        while (pos and pos.next):
            if pos in sset:
                return True

            sset.add(pos)
            pos = pos.next
        return False

print(Solution().hasCycle(ListNode(1, ListNode(2))))
